src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequestOrchestrator.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.ApiConfig.Parameters;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal.Broker;
+using Microsoft.Identity.Client.UI;
+
+namespace Microsoft.Identity.Client.Internal.Requests
+{
+
+    /// <summary>
+    /// This class decides the workflow of an interactive request. The business rules are: 
+    /// 
+    /// 1. An IWebUi based request typically handles this, if UseBroker is set to false
+    ///     1.1. Create an authorization URI
+    ///     1.2. Get an auth code. 
+    ///     1.3. Look at the code - if it's a broker install uri, then go to 2. Otherwise continue.
+    ///     1.3. Exchange the auth code for tokens
+    ///     
+    /// 2. If UseBroker is set to true or AAD requires the use of a broker, then: 
+    ///     2.1. See if a broker app can handle the request. If not, redirect to an install page.
+    ///     2.2. Otherwise, invoke the broker and wait for it to respond with a token.
+    ///     
+    ///     Question: is MSAL supposed to cache the token from the broker? (on both Android and iOS?)
+    /// </summary>
+    internal class InteractiveRequestOrchestrator
+    {
+        private readonly AuthenticationRequestParameters _requestParams;
+        private readonly AcquireTokenInteractiveParameters _interactiveParameters;
+        private readonly IServiceBundle _serviceBundle;
+
+        public InteractiveRequestOrchestrator(
+            AuthenticationRequestParameters requestParams, 
+            AcquireTokenInteractiveParameters interactiveParameters)
+        {
+            _requestParams = requestParams ?? throw new ArgumentNullException(nameof(requestParams));
+            _interactiveParameters = interactiveParameters ?? throw new ArgumentNullException(nameof(interactiveParameters));
+            _serviceBundle = requestParams.RequestContext.ServiceBundle;
+        }
+
+        // TODO: bogavril -for testing, we need to get rid of the constructor invocations here.
+        public async Task<AuthenticationResult> RunAsync(CancellationToken cancellationToken = default)
+        {
+            if (_requestParams.IsBrokerEnabled)
+            {
+                return await FetchTokensFromBrokerAsync("TODO::broker::install::url::TODO?", cancellationToken).ConfigureAwait(false);
```

Comment:
@jennyf19 - what is the redirect uri when it does not come in via the auth code?

Reply:
The redirect URI for iOS should have this format: `$"msauth.{BundleId}://auth"`

---
In reply to: [374249960](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1611#discussion_r374249960) [](ancestors = 374249960)